### ISSUE TITLE
No data in Grafana / Follow up your advise

### ISSUE THREAD

**Issue by JeromeX** [eko/pihole-exporter#32]
I just installed on my Pi4 pihole as usal, installed prometheus as well go and going through your readme. Everything is running, but there is no way to see data in grafana. Can you help me? What do i need configure on grafana site? What's wrong, what must i check?

![Bildschirmfoto 2020-03-16 um 18 02 01](https://user-images.githubusercontent.com/15369779/76782261-51a69c80-67b0-11ea-94ff-9e5a0efb2db5.png)
![Bildschirmfoto 2020-03-16 um 18 04 11](https://user-images.githubusercontent.com/15369779/76782547-b9f57e00-67b0-11ea-8b27-f8b7bb8205bb.png)

**Comment by clintcolding** [comment]
Have you added a scrape job to the prometheus.yml? Something like:

```
- job_name: pihole
  honor_timestamps: true
  metrics_path: /metrics
  scheme: http
  static_configs:
  - targets:
    - pihole-exporter.mydomain.com
```

If so, on the targets page of Prometheus, is the endpoint marked as up?

**Comment by JeromeX** [comment]
> Have you added a scrape job to the prometheus.yml? Something like:
> 
> ```
> - job_name: pihole
>   honor_timestamps: true
>   metrics_path: /metrics
>   scheme: http
>   static_configs:
>   - targets:
>     - pihole-exporter.mydomain.com
> ```

> i filled up, like on instruction page, but no success. If you have experience in this can i invite you for a remote help or something else? it might be very useful to know it must be configured step by step.  Any help is preciated.

**Comment by shaned24** [comment]
Do you have your prometheus server running on the same machine as the pi-hole-exporter?

if so then you can see your domain to `localhost` and don't forget the port `9617`
```
- job_name: pihole
  honor_timestamps: true
  metrics_path: /metrics
  scheme: http
  static_configs:
  - targets: ["localhost:9617"]
```

**Comment by rth9** [comment]
Probably the same issue:
https://github.com/eko/pihole-exporter/issues/28

**Comment by JeromeX** [comment]
Finally it's working - Yes :-)
One more question to all freaks out there... How do i run "**pihole-exporter HTTP server**" with command: **./pihole_exporter -pihole_hostname 192.168.178.2 -pihole_password** **blablabla** in the background, to work with ssh on terminal? Please send me a quickfix. Many thanks folks!

![grafik](https://user-images.githubusercontent.com/15369779/79167261-61cd8e00-7de7-11ea-9170-b6ee2d62ac28.png)
![grafik](https://user-images.githubusercontent.com/15369779/79167382-993c3a80-7de7-11ea-84cc-b9fb17e76643.png)

**Comment by rth9** [comment]
I mostly followed this:
[https://www.observability.blog/monitoring-pi-hole-with-prometheus/](url)
If you did everything else, then you need to make the service and start it.
`sudo nano /lib/systemd/system/pihole_exporter.service`
Put in this, with the appropriate paths and user:
```
[Unit]
Description=pihole_exporter

[Service]
ExecStart=/opt/pihole_exporter/pihole_exporter-linux-amd64
WorkingDirectory=/opt/pihole_exporter
Restart=always
User=pihole_exporter

[Install]
WantedBy=multi-user.target
```
And then these commands to enable it to automatically start on boot and to start it now, without having to wait for a reboot:
```
sudo systemctl daemon-reload
sudo systemctl start pihole_exporter
sudo systemctl enable pihole_exporter
```

**Comment by JeromeX** [comment]
Mission completed, many thanks 💯 , just forgot to create the service : Works perfectly on my Raspberry Pi 4. 
Last question: Can anyone tell my if is possible to add the "**_Query Log_**" into dashboard?
Right now my dashboard looks like this, it's a mix with Prometheus and Influx. Thanks JX!

![grafik](https://user-images.githubusercontent.com/15369779/79262791-1f14c000-7e92-11ea-9c81-7cc0de07aeb3.png)

**Comment by rth9** [comment]
To do a query? Or do just see the 'up to 100 queries'? Depending on your network, you will never really see anything with the number of DNS requests that are made at any given time. Plus the dashboard isn't 'live' data, like constant, scrolling DNS requests.

You could edit the dashboard, and a a link to http://pi-hole-ip-or-address/admin/queries.php to get to the page.

Maybe make a feature request because it's not one of the data points (metrics registered) that the pihole_exporter handles.

**Comment by JonAnder83** [comment]
I see a couple of issues while deploying mine.
First, I had to modify the systemd unit and add 
`ExecStart=/opt/pihole_exporter/pihole_exporter-linux-amd64 -pihole_hostnsme 192.168.20.254 -pihole_password xxxxxxxxxxx`

Then, the dashboard import from grafana web is different than the one here, in GitHub. While this one lets me select Prometheus as the data source, I still get no data on my dashboard.
Any help?
Sorry for piggybacking on this post but I didn't see it fit to open yet another thread with the same issue.
Thanks!

**Comment by OfficialTehlo** [comment]
I'm having a similar issue to you @JonAnder83 however, after adding the rule you had, I got the following:

`May 05 07:07:35 plex-server pihole_exporter-linux-amd64[21615]: 2020/05/05 07:07:35 Unable to unmarshal PI-Hole statistics to statistics struct model invalid character 'F' looking for beginning of value
May 05 07:07:35 plex-server pihole_exporter-linux-amd64[21615]: 2020/05/05 07:07:35 New tick of statistics: 0 ads blocked / 0 total DNS queries
May 05 07:07:45 plex-server pihole_exporter-linux-amd64[21615]: 2020/05/05 07:07:45 Unable to unmarshal PI-Hole statistics to statistics struct model invalid character 'F' looking for beginning of value
May 05 07:07:45 plex-server pihole_exporter-linux-amd64[21615]: 2020/05/05 07:07:45 New tick of statistics: 0 ads blocked / 0 total DNS queries
May 05 07:07:55 plex-server pihole_exporter-linux-amd64[21615]: 2020/05/05 07:07:55 Unable to unmarshal PI-Hole statistics to statistics struct model invalid character 'F' looking for beginning of value
May 05 07:07:55 plex-server pihole_exporter-linux-amd64[21615]: 2020/05/05 07:07:55 New tick of statistics: 0 ads blocked / 0 total DNS queries
May 05 07:08:05 plex-server pihole_exporter-linux-amd64[21615]: 2020/05/05 07:08:05 Unable to unmarshal PI-Hole statistics to statistics struct model invalid character 'F' looking for beginning of value
May 05 07:08:05 plex-server pihole_exporter-linux-amd64[21615]: 2020/05/05 07:08:05 New tick of statistics: 0 ads blocked / 0 total DNS queries
May 05 07:08:15 plex-server pihole_exporter-linux-amd64[21615]: 2020/05/05 07:08:15 Unable to unmarshal PI-Hole statistics to statistics struct model invalid character 'F' looking for beginning of value
May 05 07:08:15 plex-server pihole_exporter-linux-amd64[21615]: 2020/05/05 07:08:15 New tick of statistics: 0 ads blocked / 0 total DNS queries`

Any idea on how to fix this?
Thanks :)

**Comment by rth9** [comment]
@JonAnder83 did you edit each graph to be Prometheus?

@OfficialTehlo what hardware are you running? On a Pi itself? In a VM? Make sure the pihole_exporter is for your hardware. Did you follow the steps in the observability.blog link above with creating the user, etc.? The ExecStart line needs to be customised for your environment with the right executable, the right host, and the right password.

**Comment by OfficialTehlo** [comment]
Hi @rth9 yes I have, because when I run the command by itself, then I get the following:

![image](https://user-images.githubusercontent.com/51698749/81082078-104e8580-8ef3-11ea-9a11-a1d7b29faffe.png)

So on its own it's working fine, but in a systemd it isn't.

**Comment by rth9** [comment]
You are using the exact same command to start it manually as you put in the ExecStart line? Issue [#26](https://github.com/eko/pihole-exporter/issues/26) shows the same error?

**Comment by JonAnder83** [comment]
> 
> 
> @JonAnder83 did you edit each graph to be Prometheus?
> 

Nope. Funny thing: If you have multiple data sources (i.e. influxdb and prometheus) and set prometheus as the default data source, the import just works out of the box.
If other data source, i.e.: grafana is your main data source while importing, you will have to re-select data source editing all the elements of the dashboard and then, format them appropriately.

Thanks though!

**Comment by OfficialTehlo** [comment]
@rth9 Yep, same command but still have the same issue. Any idea?

**Comment by magefesa** [comment]
> Have you added a scrape job to the prometheus.yml

How can i do that on the docker ?Thanks

**Comment by 91ajames** [comment]
Tried this once before and even today getting back to it, I'm still lost on how to make it show up in the graph.
Tried a few things which still didn't work and then all of a sudden, it works but not the way it should work.
Just banjo green graphs and statistics that doesn't mean anything.

Using Docker-Compose in case anyone is wondering.
p 9617 in web view, 404 page not found

I have it working on Prometheus,

Tried using _./pihole_exporter -pihole_hostname 192.168.178.2 -pihole_password blablabla_ 
but no idea where' im placing that cmd in.

Can't bash into Ekofr-Pi-Hole

**Comment by rth9** [comment]
If this works (make sure to change path to your executable and if it's not AMD64), I put single quote around my password:
`/opt/pihole_exporter/pihole_exporter-linux-amd64 -pihole_hostname IP_ADDRESS -pihole_password 'PASSWORD'`

Then you can do the following should work to get it as a service.

File path/name:
`/lib/systemd/system/pihole_exporter.service`

I prefer nano, so:
`sudo nano /lib/systemd/system/pihole_exporter.service`

And put in these contents (mine is AMD64, your file may be different architecture or location), it requires that you created the "pihole_exporter" user account manually.:
> [Unit]
> Description=pihole_exporter
> 
> [Service]
> ExecStart=/opt/pihole_exporter/pihole_exporter-linux-amd64 -pihole_hostname IP_ADDRESS -pihole_password 'PASSWORD'
> WorkingDirectory=/opt/pihole_exporter
> Restart=always
> User=pihole_exporter
> 
> [Install]
> WantedBy=multi-user.target

Command to enable it as a service and auto-start on boot:
`sudo systemctl enable pihole_exporter`

You can then start it with this if you don't want to reboot:
`sudo systemctl start pihole_exporter`

Looking at the status of the service or using journalctl should show essentially the same output as running the command manually. Don't run the command manually when the service is started.

**Comment by 91ajames** [comment]
Actually, somehow when i looked into it on where the graph was coming from. it was from Grafana itself and not from the Prometheus.
![pihole-export](https://user-images.githubusercontent.com/24540531/92270929-e6af6600-eeb4-11ea-9fe7-b464bf5601cc.png)

- docker compose/ armfh

**Comment by rth9** [comment]
So you are using the ARM binary? Where my example above has amd64, you have replaced the commands to point to the arm binary? Your screenshot has a tick beside Prometheus, not Grafana, are you sure it's looking at the wrong place? I ran into a problem where my default was InfluxDB, I added Prometheus after, but had to update my dashboard to point to Prometheus instead of InfluxDB. Are you saying your default datasource was Grafana instead of InfluxDB or Prometheus?

**Comment by 91ajames** [comment]
I had just recently added InfluxDB to try it out which im not getting much out of it.
the 1st pic was under default. Not sure how it got bumped there so i had put it back to Prometheus which resulted no graph.
I'de assumed default was grafana itself resulted the green graph pic.

this is where im at. Tottaly ignoring InfluxDB
![pihole-export](https://user-images.githubusercontent.com/24540531/92273009-9c2fe880-eeb8-11ea-9409-9da861174d0f.png)

**Comment by rth9** [comment]
So you have data. I ended up going to edit the dashboard, not each graph, then JSON Model, copy and paste out, find and replace where it said "datasource": "default", and replaced it with "datasource": "Prometheus", and then copied and pasted it back in. That cleared up my graphs showing no or wrong data.

**Comment by 91ajames** [comment]
ok.  Changed one and looks like it replaced all of them for me except _"datasource": null_ , which I had also replaced.
https://pastebin.com/BpjUVzeD

**Comment by rth9** [comment]
Mine is still null and mine is also 9311.

Your dashboard now is filled with graphs with no data? Not odd stuff like with InfluxDB or Grafana previously?

**Comment by 91ajames** [comment]
Its just blank.
https://streamable.com/cfdd89

Which node?

**Comment by lkshck** [comment]
> @JeromeX Do you have your prometheus server running on the same machine as the pi-hole-exporter?
> 
> if this is the case then you need to set your domain to `localhost` and don't forget the port `9617`
> 
> ```
> - job_name: pihole
>   honor_timestamps: true
>   metrics_path: /metrics
>   scheme: http
>   static_configs:
>   - targets: ["localhost:9617"]
> ```

Where can i find the prometheus.yml to update the target?

**Comment by JeromeX** [comment]
Sure, it work’s meanwhile fantastic and smooth, thanks for the info 
    	
    	

    	 Holen Sie sich Outlook für iOS
    
  




On Sat, Sep 5, 2020 at 8:36 AM +0200, "lkshck" <notifications@github.com> wrote:
















@JeromeX Do you have your prometheus server running on the same machine as the pi-hole-exporter?


if this is the case then you need to set your domain to localhost and don't forget the port 9617
- job_name: pihole
  honor_timestamps: true
  metrics_path: /metrics
  scheme: http
  static_configs:
  - targets: ["localhost:9617"]




Where can i find the prometheus.yml to update the target?

**Comment by TehloWasTaken** [comment]
> > @JeromeX Do you have your prometheus server running on the same machine as the pi-hole-exporter?
> > 
> > if this is the case then you need to set your domain to `localhost` and don't forget the port `9617`
> > 
> > ```
> > - job_name: pihole
> >   honor_timestamps: true
> >   metrics_path: /metrics
> >   scheme: http
> >   static_configs:
> >   - targets: ["localhost:9617"]
> > ```
> 
> Where can i find the prometheus.yml to update the target?

Can find this in `/etc/prometheus/prometheus.yml` on your typical install.

**Comment by lkshck** [comment]
> > Where can i find the prometheus.yml to update the target?
> 
> Can find this in `/etc/prometheus/prometheus.yml` on your typical install.

I'm running the exporter as an docker image. So i'm not able to find the yml under this path. Is it somewhere else located in this case?

**Comment by 91ajames** [comment]
@lkshck 
I had to locate the yml file online since on my end in my docker, it couldn't create it for some reason.
After moving it and restarting it some how created another file I believe but didn't look into it before I had deleted it, oops.
but basically, u need to draw a path from host to path in docker of the oppose file in the container to make adjustments.


but with my issue not showing up, think its got to due with the network system. Ekofr is on host, graf+prometheus is on bridge. Even after editing from 0.0.0.0/localhost to the attach host network of 172.17.0.1, which still sees it but still doesn't show any graph status in grafana. can't attach those two to host for some reason. I keep running into
Failure
container cannot be disconnected from host network or connected to host network

**Comment by lkshck** [comment]
> @lkshck
> I had to locate the yml file online since on my end in my docker, it couldn't create it for some reason.
> After moving it and restarting it some how created another file I believe but didn't look into it before I had deleted it, oops.
> but basically, u need to draw a path from host to path in docker of the oppose file in the container to make adjustments.
> 
> @rth9
> After moving stuff to host to make sure everything is good, decided to goto the other pihole_exporter-1, part of the field has been inputing graph, the rest blank, took part of the code metrics and paste it into Pi-Hole Exporter and Boom, its there, sucks having to put everything by manual instead of auto set. Metrics are left blank.
> 
> ![piholeaaaaa](https://user-images.githubusercontent.com/24540531/92314886-22792700-efac-11ea-8d45-62bf12f93b87.png)
> 
> Honestly, this could be alot better tho for the default without putting in so much tweaking into it. and the error of % ads blocked "O.toFixed is not a function" when selecting Premethuis
> (selected a dupe for Domain being blocked)
> ![piholeaaaaa](https://user-images.githubusercontent.com/24540531/92315012-c0b9bc80-efad-11ea-8b21-92874fc4f2c8.png)

Using now the typical installation of Prometheus and target looks ok from my point of view, but still HTTP Error Not Found. Any other ideas?
<img width="1144" alt="Bildschirmfoto 2020-09-07 um 20 27 44" src="https://user-images.githubusercontent.com/46962619/92411572-af58e780-f148-11ea-8521-c8e626696d40.png">
<img width="655" alt="Bildschirmfoto 2020-09-07 um 20 28 01" src="https://user-images.githubusercontent.com/46962619/92411578-b1bb4180-f148-11ea-8ba7-d0b66d1a330e.png">

**Comment by TehloWasTaken** [comment]
@lkshck you don't want to add the `/metrics` to your datasource setup. Just the http://localhost:9090 is enough for it to draw the data out of Prometheus.

**Comment by lkshck** [comment]
@TehloWasTaken did that and i was able to save the datasource and dashboard is showing some data but not all of them. I thought i need to define http://localhost:9617 as i defined this in the prometheus job. Dashboard now looks like the following:
![grafik](https://user-images.githubusercontent.com/46962619/92414990-2d23ef80-f157-11ea-93a8-463889c23d41.png)

**Comment by TehloWasTaken** [comment]
@lkshck Looks like it's getting there! I'd make sure the missing metrics are in Prometheus, if they are not they will simply not be exported from your PiHole.

Also, holy crap you block a lot of domains. You do know that more != better for PiHole right? 😛

**Comment by lkshck** [comment]
@TehloWasTaken Yes I'm aware of that 😆 didn't found the best list setup for now. 

As i'm not that familiar with Prometheus. Any way to check if the is in Prometheus? Like Top queries by domain..

**Comment by 91ajames** [comment]
That's funny.
![pipipipip](https://user-images.githubusercontent.com/24540531/92416436-778f7700-f12b-11ea-9234-b2b8dd097588.png)

**Comment by steventblack** [comment]
I think there may be a missing statement in the pihole-exporter.json that's causing a lot of this frustration. Poking around a bit, I noticed that it makes heavy use of the variable ${DS_PROMETHEUS} in the template but doesn't actually provide the definition for it. I amended the block to include a definition for it and everything snapped into place without having to edit any of the panels.

This is the block I added:

```
{
    "current": {},
    "includeAll": false,
    "label": "Datasource",
    "name": "DS_PROMETHEUS",
    "options": [],
    "query": "prometheus",
    "refresh": 1,
    "regex": "",
     "type": "datasource"
},
```

I added it as the first item in the list for the "templating" block (near the very end of the file).
